java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,351 @@
+    this.userPrivateKey = userPrivateKey;
+    initMacKeys(peerKey, minPeerTagLength);
+  }
+
+  public XECPublicKey peerKey() {
+    return peerKey;
+  }
+
+  public XECPublicKey userPublicKey() {
+    return userPublicKey;
+  }
+
+  public XECPrivateKey userPrivateKey() {
+    return userPrivateKey;
+  }
+
+  /**
+   * Generates a user's tag corresponding to the given message {@code msg} and the counter {@code
+   * cnt}.
+   *
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   */
+  public byte[] generateTag(byte[] msg, int cnt) throws CounterOutOfBoundsException {
+    return generateTag(msg, cnt, this.userMacKey);
+  }
+
+  /**
+   * Checks whether the peer's tag matches received message and some counter.
+   *
+   * @param peerTag tag from a peer.
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the peer to the user.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than {@code MAX_TAG_LENGTH}) or it's not equal to the prefix of a
+   * correct tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          String.format("The received tag has invalid length. Expected %d or more, got %d.",
+              minPeerTagLength, peerTag.length));
+    }
+
+    byte[] truncatedTag = Arrays.copyOf(generateTag(msg, cnt, this.peerMacKey), peerTag.length);
+    if (!Arrays.equals(peerTag, truncatedTag)) {
+      throw new WrongTagException("The received tag doesn't match the expected tag.");
+    }
+  }
+
+  /**
+   * Common part of the constructors. Initialises peer's public key, minimum length of a peer's tag,
+   * base point for Curve25519. Calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum length of a peer's tag.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  private void initMacKeys(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException, InvalidKeyException {
+    this.peerKey = peerKey;
+    if (minPeerTagLength < MIN_TAG_LENGTH || minPeerTagLength > MAX_TAG_LENGTH) {
+      throw new MinPeerTagLengthOutOfBoundsException(
+          String.format("minPeerTagLength argument should be in [1..32] range. Got %d.",
+              minPeerTagLength));
+    }
+    this.minPeerTagLength = minPeerTagLength;
+
+    KeyAgreement ka;
+    try {
+      ka = KeyAgreement.getInstance("XDH");
+    } catch (NoSuchAlgorithmException e) {
```

Comment:
I can't seem to add to the old conversation around asserting this at module initialization time. So re-reading the code I am a little upset about the construction again. KeyAgreement seems to carry state and getInstance does not return a factory to construct new KeyAgreements. In that case I'd mostly add a static initializer block if Java let's you do this that asserts that XDH is actually registered:

```java
static {
  KeyAgreement.getInstance("XDH");
}
```

Hopefully that'd fail startup if this precondition is violated.

Reply:
Made an init block instead of static one (otherwise there will be just one Mac object for all the Glome objects).